src/utils/daos_hdlr.c

```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	} else {
+		*_cont_name = strdup(cont_name);
+		if (*_cont_name == NULL) {
+			rc = -ENOMEM;
+			goto out;
+		}
+	}
+	*_obj_name = strdup(fname);
+	if (*_obj_name == NULL) {
+		free(*_cont_name);
+		*_cont_name = NULL;
+		rc = -ENOMEM;
+		goto out;
+	}
+	out:
+		if (f1)
+			free(f1);
+		if (f2)
+			free(f2);
+	return rc;
+}
+
+static ssize_t
+daos_write(struct daos_file_t *daos_file, char *file, void *buf, size_t size)
+{
+	int rc;
+	/* record address and size of user buffer in io vector */
+	d_iov_t iov;
+	d_iov_set(&iov, buf, size);
+
+	/* define scatter-gather list for dfs_write */
+	d_sg_list_t sgl;
+	sgl.sg_nr = 1;
+	sgl.sg_iovs = &iov;
+	sgl.sg_nr_out = 1;
+
+	rc = dfs_write(daos_file->dfs, daos_file->obj, &sgl, daos_file->offset, NULL); 
+
+	if (rc) {
+		fprintf(stderr, "dfs_write %s failed (%d %s)\n",
+			file, rc, strerror(rc));
+		errno = rc;
+		size = -1;
+	}
+
+	daos_file->offset += (daos_off_t)size;
+	return (ssize_t) size;
+}
+
+static ssize_t
+daos_file_write(daos_file_t *daos_file, char* file,
+		void* buf, size_t size)
+{
+	ssize_t num_bytes_written = 0;
+	if (daos_file->type == POSIX) {
+		num_bytes_written = write(daos_file->fd, buf, size);
+	} else if (daos_file->type == DAOS) {
+		num_bytes_written = daos_write(daos_file, file, buf, size);
+	} else {
+		fprintf(stderr, "File type not known: %s type=%d\n",
+			file, daos_file->type);
+	}
+	if (num_bytes_written < 0) {
+		fprintf(stderr, "write error on %s type=%d\n",
+			file, daos_file->type);
+	}
+	return num_bytes_written;
+}
+
+static void 
+daos_open(daos_file_t *daos_file, char *file, int flags, mode_t mode)
+{
+
+	int rc; 
+	dfs_obj_t *parent = NULL;
+	char *name = NULL, *dir_name = NULL;
+
+	parse_filename(file, &name, &dir_name);
+
+	assert(dir_name);
+
+	rc = dfs_lookup(daos_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
+	if (parent == NULL) {
+		fprintf(stderr, "dfs_lookup %s failed \n", dir_name);
+	}
+
+	rc = dfs_open(daos_file->dfs, parent, name, mode | S_IFREG,
+		flags, 0, 0, NULL, &(daos_file->obj));
+	if (rc)
+		fprintf(stderr, "dfs_open %s failed (%d)\n", name, rc);
+}
+
+void
+daos_file_open(daos_file_t *daos_file, char* file, int flags, ...)
+{
+	/* extract the mode */
+	int mode_set = 0;
+	mode_t mode  = 0;
+	if (flags & O_CREAT) {
+		va_list ap;
+		va_start(ap, flags);
+		mode = va_arg(ap, mode_t);
+		va_end(ap);
+		mode_set = 1;
+	}
+
+	if (daos_file->type == POSIX) {
+		if (mode_set) {
+			daos_file->fd = open(file, flags, mode);
+		} else {
+			daos_file->fd = open(file, flags);
+		}
+	} else if (daos_file->type == DAOS) {
+		daos_open(daos_file, file, flags, mode);
+	} else {
+		fprintf(stderr, "File type not known: %s type=%d\n",
+			file, daos_file->type);
+	}
+}
+
+static int
+daos_mkdir(daos_file_t *daos_file, const char *path)
+{
+	int rc = 0;
+	dfs_obj_t *parent = NULL;
+	char *name = NULL;
+	char *dname = NULL;
+	parse_filename(path, &name, &dname);
+    
+	assert(dname);
+	if (name && strcmp(name, "/") != 0) {
```

Comment:
i don't understand this if condition.
how can name be "/" in this case?
and if name is NULL, why do we not return an error??

Reply:
I think this is carry-over from mpifileutils trying to create the directory "/" in DFS.
Maybe we don't need that here, since this doesn't work exactly the same way as mpifileutils.